[TT-11929, DX-1279] Clarification of Go template behaviour - response transform (#4487)
TykTechnologies/tyk-docs: tyk-docs/content/advanced-configuration/transform-traffic/response-body.md

- Inbound form or query data can be accessed through the `._tyk_context.request_data` namespace where it will be available in as a `key:[]value` map
- values from [key-value (KV) storage]({{< ref "tyk-configuration-reference/kv-store#transformation-middleware" >}}) can be injected into the template using the notation appropriate to the location of the KV store
 

> The response body transform middleware can iterate through list indices in dynamic data so, for example, calling `{{ index ._tyk_context.request_data.variablename 0 }}` in a template will expose the first entry in the `request_data.variablename` key/value array.

Reviewer: Why is only the first entry in the key/value array and not the other list indices in the array?

Author: Same question as the request body PR, same answer :)

{{< note success >}}

>  **Note**  

Reviewer 2: ```suggestion
**Note**  
```

Whitespace

> As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. This data structure consists of the request context variables (`_tyk_context`) and session metadata (`_tyk_meta`) - if enabled for the middleware - concatenated with the response body.

Reviewer 2: ```suggestion
As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. This data structure consists of the request context variables (`_tyk_context`) and session metadata (`_tyk_meta`) - if enabled for the middleware - together with the response body.
```

> As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. The template receives the decoded JSON or XML of the request body. If session variables or meta data are enabled, additional fields will be provided: `_tyk_context` and `_tyk_meta` respectively.

Author: ```suggestion
As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. The template receives the decoded JSON or XML of the response body. If session variables or meta data are enabled, additional fields will be provided: `_tyk_context` and `_tyk_meta` respectively.
```

>  As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided with the response body in the data structure to which the template is applied and so will be processed accordingly if, for example, the template applies a transformation to the entire data structure.

Reviewer 2: ```suggestion
As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. As the template applies a transformation to the entire data structure it receives, context variables and session metadata will be available if enabled, along with the response body. Otherwise, just the response body is available.
```

While technically correct, I found the second sentence nearly impossible to consume. I tried breaking apart into a few sentences to be easier to read, but if you can reword it, please.

Reviewer: @andyo-tyk @titpetric 

```suggestion
As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided with the response body in the data structure to which the template is applied. Subsequently, these will be merged if the template applies a transformation to the entire data structure.
```

>  As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided with the response body in the data structure to which the template is applied and so will be processed accordingly if, for example, the template applies a transformation to the entire data structure.

Author: ```suggestion
As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. This data structure consists of the request context variables (`_tyk_context`) and session metadata (`_tyk_meta`) - if enabled for the middleware - concatenated with the response body.
```

>  As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided alongside the response body in the data structure to which the template is applied and so will be processed accordingly if, for example, the template applies a transformation to the entire data structure.

Reviewer: ```suggestion
As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided with the response body in the data structure to which the template is applied and will be merged if the template applies a transformation to the entire data structure.
```

>  As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided alongside the response body in the data structure to which the template is applied and so will be processed accordingly if, for example, the template applies a transformation to the entire data structure.

Author: ```suggestion
 As explained in the [documentation](https://pkg.go.dev/text/template), templates are executed by applying them to a data structure. Context variables and session metadata are, if enabled, provided with the response body in the data structure to which the template is applied and so will be processed accordingly if, for example, the template applies a transformation to the entire data structure.
```
